In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import warnings
warnings.filterwarnings('ignore')
pd.set_option("display.max_rows", None)
pd.set_option('display.float_format', '{:.2f}'.format)

In [3]:
file ='house-prices.csv'
df = pd.read_csv(file)
df.drop(['Order', 'PID'], axis=1, inplace=True)

In [4]:
df.head()

MS SubClass MS Zoning  Lot Frontage  Lot Area Street Alley Lot Shape  \
0           60        RL           NaN      8795   Pave   NaN       IR1   
1           20        RL         75.00     10170   Pave   NaN       Reg   
2          160        RM         21.00      2001   Pave   NaN       Reg   
3           20        RL         70.00     10552   Pave   NaN       IR1   
4          190        RM         60.00     10120   Pave   NaN       IR1   

  Land Contour Utilities Lot Config  ... Pool Area Pool QC  Fence  \
0          Lvl    AllPub     Inside  ...         0     NaN    NaN   
1          Lvl    AllPub     Corner  ...         0     NaN    NaN   
2          Lvl    AllPub     Inside  ...         0     NaN    NaN   
3          Lvl    AllPub     Inside  ...         0     NaN    NaN   
4          Bnk    AllPub     Inside  ...         0     NaN  MnPrv   

  Misc Feature Misc Val Mo Sold  Yr Sold  Sale Type  Sale Condition  SalePrice  
0          NaN        0       4     2009        WD           Normal     236000  
1          NaN        0       6     2006        WD           Normal     155000  
2          NaN        0       1     2007        WD           Normal      75000  
3          NaN        0       4     2010        WD           Normal     165500  
4          NaN        0       1     2007        WD           Normal     122000  

[5 rows x 80 columns]

In [5]:
df.shape

(2430, 80)

__Observation__:
- There are 80 features and 2430 entries
- We observe there's a mix of categorical and numerical features

Data types of the different columns in the dataset:

In [6]:
df.dtypes.unique()

array([dtype('int64'), dtype('O'), dtype('float64')], dtype=object)

In [7]:
continuous_columns = df.dtypes[df.dtypes == 'float64'].index
continuous_columns

Index(['Lot Frontage', 'Mas Vnr Area', 'BsmtFin SF 1', 'BsmtFin SF 2',
       'Bsmt Unf SF', 'Total Bsmt SF', 'Bsmt Full Bath', 'Bsmt Half Bath',
       'Garage Yr Blt', 'Garage Cars', 'Garage Area'],
      dtype='object')

In [8]:
discrete_columns = df.dtypes[df.dtypes == 'int64'].index
discrete_columns

Index(['MS SubClass', 'Lot Area', 'Overall Qual', 'Overall Cond', 'Year Built',
       'Year Remod/Add', '1st Flr SF', '2nd Flr SF', 'Low Qual Fin SF',
       'Gr Liv Area', 'Full Bath', 'Half Bath', 'Bedroom AbvGr',
       'Kitchen AbvGr', 'TotRms AbvGrd', 'Fireplaces', 'Wood Deck SF',
       'Open Porch SF', 'Enclosed Porch', '3Ssn Porch', 'Screen Porch',
       'Pool Area', 'Misc Val', 'Mo Sold', 'Yr Sold', 'SalePrice'],
      dtype='object')

In [9]:
categorical_columns = df.dtypes[df.dtypes == 'object'].index
categorical_columns

Index(['MS Zoning', 'Street', 'Alley', 'Lot Shape', 'Land Contour',
       'Utilities', 'Lot Config', 'Land Slope', 'Neighborhood', 'Condition 1',
       'Condition 2', 'Bldg Type', 'House Style', 'Roof Style', 'Roof Matl',
       'Exterior 1st', 'Exterior 2nd', 'Mas Vnr Type', 'Exter Qual',
       'Exter Cond', 'Foundation', 'Bsmt Qual', 'Bsmt Cond', 'Bsmt Exposure',
       'BsmtFin Type 1', 'BsmtFin Type 2', 'Heating', 'Heating QC',
       'Central Air', 'Electrical', 'Kitchen Qual', 'Functional',
       'Fireplace Qu', 'Garage Type', 'Garage Finish', 'Garage Qual',
       'Garage Cond', 'Paved Drive', 'Pool QC', 'Fence', 'Misc Feature',
       'Sale Type', 'Sale Condition'],
      dtype='object')

__Observation__:
- Comparing the data types with the documentation about the dataset we observe the following mismatches:
- Variables that should be categorical: __'MS SubClass'__;
- Variables that should be continuous : __'Lot Area', '1st Flr SF', '2nd Flr SF', '1st Flr SF', '2nd Flr SF', 'Low Qual Fin SF', 'Gr Liv Area', 'Wood Deck SF', 'Open Porch SF', 'Enclosed Porch','3Ssn Porch', 'Screen Porch', 'Pool Area', 'Misc Val'and 'SalePrice'__;
- Variables that should be discrete   : __'Bsmt Full Bath', 'Bsmt Half Bath', 'Garage Yr Blt', 'Garage Cars'__;
- Observing the values of the variables that should be discrete we see that see take only integer values as expected;
  
__Solution__:
- We convert the datatypes of all the above variables to the correct ones;
- For the discrete variables (`int64`) we first convert all the variables to `Int64`, since only this integer datatype supports `NA` values;

In [10]:
# Convert to Categorical
df['MS SubClass'] = df['MS SubClass'].astype(str)

# Convert to Discrete (but first convert all int64 to Int64 )
for col in discrete_columns:
        df[col] = df[col].astype('Int64')

# Convert to Continuous
convert_to_continuous = ['Lot Area', '1st Flr SF', '2nd Flr SF', '1st Flr SF', '2nd Flr SF', 'Low Qual Fin SF', 'Gr Liv Area', 'Wood Deck SF', 'Open Porch SF', 'Enclosed Porch','3Ssn Porch', 'Screen Porch', 'Pool Area', 'Misc Val', 'SalePrice']
for col in convert_to_continuous:
        df[col] = df[col].astype('float64')


convert_to_discrete = ['Bsmt Full Bath', 'Bsmt Half Bath', 'Garage Yr Blt', 'Garage Cars']
for col in convert_to_discrete:
        df[col] = df[col].astype('Int64')


We obtain again the list of datatypes after the conversions:

In [11]:
continuous_columns = df.dtypes[df.dtypes == 'float64'].index
discrete_columns = df.dtypes[df.dtypes == 'Int64'].index
categorical_columns = df.dtypes[df.dtypes == 'object'].index

## Categorical Features

In [12]:
print('Number of categorical features: ', {len(categorical_columns)})

Number of categorical features:  {43}


### Missing values:

In [13]:
print(f'Total missing values: {df[categorical_columns].isna().sum().sum()}')

Total missing values: 12524


In [20]:
df_categorical_missing = df[categorical_columns].isna().sum().to_frame(name='total')
df_categorical_missing = df_categorical_missing[df_categorical_missing['total']!=0].sort_values('total', ascending=False)
df_categorical_missing['percentage'] = df_categorical_missing['total']/len(df)*100
df_categorical_missing

Empty DataFrame
Columns: [total, percentage]
Index: []

__Observation__:
- Some of the categorical features present a very high percentage of missing values

__Solution__:
- Discard all features with missing values >75%
- Obtain again the list of categorical features after the column removal:

In [15]:
df.drop(df_categorical_missing[df_categorical_missing['percentage']>75].index, axis=1)
categorical_columns = df.dtypes[df.dtypes == 'object'].index

__Observation__:
- For many of the categorical ordinal features the label 'NA' (indicating that a given feature is not present in the property, e.g. a basement) was converted to a missing value in dataframe.


__Solution__:
- Replace the missing `np.nan` values of all these features (except 'Electrical') with the string 'NA'

In [17]:
categorical_missing = ['Alley', 'Mas Vnr Type', 'Bsmt Qual', 'Bsmt Cond', 'Bsmt Exposure','BsmtFin Type 1', 'BsmtFin Type 2', 'Fireplace Qu', 'Garage Type', 'Garage Finish', 'Garage Qual', 'Garage Cond', 'Pool QC', 'Fence', 'Misc Feature']
df[categorical_missing] = df[categorical_missing].fillna('NA')

In [18]:
df[df['Electrical'].isna()]


MS SubClass MS Zoning  Lot Frontage  Lot Area Street Alley Lot Shape  \
694           80        RL         73.00   9735.00   Pave    NA       Reg   

    Land Contour Utilities Lot Config  ... Pool Area Pool QC Fence  \
694          Lvl    AllPub     Inside  ...      0.00      NA    NA   

    Misc Feature Misc Val Mo Sold  Yr Sold  Sale Type  Sale Condition  \
694           NA     0.00       5     2008        WD           Normal   

     SalePrice  
694  167500.00  

[1 rows x 80 columns]

__Observation__:
- The only categorical feature with remaining missing values is the 'Electrical' feature, with a single entry

__Solution__:
- Remove this entry

In [19]:
df.drop(df[df['Electrical'].isna()].index, inplace=True)

## Numerical discrete columns

In [20]:
print('Number of numerical discrete features:', len(discrete_columns))

Number of numerical discrete features: 17


### Missing Values:

In [21]:
print(f'Total missing values: {df[discrete_columns].isna().sum().sum()}')

Total missing values: 143


In [21]:
df_discrete_missing = df[discrete_columns].isna().sum().to_frame(name='total')
df_discrete_missing = df_discrete_missing[df_discrete_missing['total']!=0].sort_values('total', ascending=False)
df_discrete_missing['percentage'] = df_discrete_missing['total']/len(df)*100
df_discrete_missing

total  percentage
Garage Yr Blt     138        5.68
Bsmt Full Bath      2        0.08
Bsmt Half Bath      2        0.08
Garage Cars         1        0.04

__Observation__:
- 'Garage Yr Blt' is the feature with more missing values, these correspond to entries where there is no garage.
- Other features represent very small number of missing values and will be handled in Section - Inconcistencies
__Solution__: 
- For 'Garage Yr Blt': 
    - check if the missing values correpond mostly to propreties where a garage is not present;
    - if yes, it's not obvious which value to give to 'Garage Yr Blt' these cases...
    - check correlation between 'Year Blt' and 'Garage Yr Blt'
    - if correlation is high we decide to drop the 'Garage Yr Blt feature'

In [22]:
df[df['Garage Yr Blt'].isna()]['Garage Type'].value_counts()

Garage Type
NA        136
Detchd      2
Name: count, dtype: int64

In [23]:
df['Garage Yr Blt'].corr(df['Year Built'])

0.8354174366616421

As explained before, we decide to drop the feature 'Garage Yr Blt'

In [24]:
df.drop('Garage Yr Blt', axis=1, inplace=True)

# Numerical continuous variable

In [25]:
print('Number of numerical continuous features:', len(continuous_columns))

Number of numerical continuous features: 20


### Missing Values:

In [26]:
print(f'Total missing values: {df[continuous_columns].isna().sum().sum()}')

Total missing values: 445


In [44]:
df_continuous_missing = df[continuous_columns].isna().sum().to_frame(name='total')
df_continuous_missing = df_continuous_missing[df_continuous_missing['total']!=0].sort_values('total', ascending=False)
df_continuous_missing['percentage'] = df_continuous_missing['total']/len(df)*100
df_continuous_missing

total  percentage
Garage Area      1        0.04

__Observation__:
- Feature 'Lot Frontage' has a significant ammount of missing values, these could be from properties with no street connection, but this should be verified
- 'Mas Vnr Area' (Masonry veneer area in square feet) with missing values could also be coming from properties with no Masonry veneer surface

__Solution__:
- Observe relation between 'Lot Frontage' and 'Lot Config' for properties with no lot frontage.
- Check if missing values in 'Mas Vnr Area', also have 'Mas Vnr Type' equal to 'NA'
- If previous 

In [28]:
df[df['Lot Frontage'].isna()]['Lot Config'].value_counts()

Lot Config
Inside     235
Corner      86
CulDSac     79
FR2         17
FR3          3
Name: count, dtype: int64

Most missing values in 'Lot Frontage' correspond to lot configurations where it's more likely that there's no connection to street. Hence it seemes reasonable that these missing values should correpond to 0 feet of street connection of the property.

In [29]:
df['Lot Frontage'].fillna(0, inplace=True)

In [30]:
df[df['Mas Vnr Area'].isna()]['Mas Vnr Type'].value_counts()

Mas Vnr Type
NA    20
Name: count, dtype: int64

All missing values in 'Mas Vnr Area' correspond to entries where 'Mas Vnr Type' is 'NA', hence it makes sense that these values should actually be zero.

In [31]:
df['Mas Vnr Area'].fillna(0, inplace=True)

Next we look at the other missing values in some Basement related features. Below we see that the 3 missing values correspond to a single entry, and we decide to drop this entry.

In [38]:
df[df['BsmtFin SF 1'].isna() | df['BsmtFin SF 2'].isna() | df['Bsmt Unf SF'].isna()]

Empty DataFrame
Columns: [MS SubClass, MS Zoning, Lot Frontage, Lot Area, Street, Alley, Lot Shape, Land Contour, Utilities, Lot Config, Land Slope, Neighborhood, Condition 1, Condition 2, Bldg Type, House Style, Overall Qual, Overall Cond, Year Built, Year Remod/Add, Roof Style, Roof Matl, Exterior 1st, Exterior 2nd, Mas Vnr Type, Mas Vnr Area, Exter Qual, Exter Cond, Foundation, Bsmt Qual, Bsmt Cond, Bsmt Exposure, BsmtFin Type 1, BsmtFin SF 1, BsmtFin Type 2, BsmtFin SF 2, Bsmt Unf SF, Total Bsmt SF, Heating, Heating QC, Central Air, Electrical, 1st Flr SF, 2nd Flr SF, Low Qual Fin SF, Gr Liv Area, Bsmt Full Bath, Bsmt Half Bath, Full Bath, Half Bath, Bedroom AbvGr, Kitchen AbvGr, Kitchen Qual, TotRms AbvGrd, Functional, Fireplaces, Fireplace Qu, Garage Type, Garage Finish, Garage Cars, Garage Area, Garage Qual, Garage Cond, Paved Drive, Wood Deck SF, Open Porch SF, Enclosed Porch, 3Ssn Porch, Screen Porch, Pool Area, Pool QC, Fence, Misc Feature, Misc Val, Mo Sold, Yr Sold, Sale Type, Sale Condition, SalePrice]
Index: []

[0 rows x 79 columns]

In [33]:
df.drop([104], inplace=True)

We still have one missing value in the 'Garage Surface' feature. Let's see what happens with this entry.

# Inconsistencies

In [ ]:
df_missing = df.isna().sum().to_frame(name='total')
df_missing = df_missing[df_missing['total']!=0].sort_values('total', ascending=False)
df_missing['percentage'] = df_missing['total']/len(df)*100
df_missing.head()

### Basement variables

Next we check for possible inconsistencies in Basement related features

In [ ]:
basement_columns = ['Bsmt Qual', 'Bsmt Cond', 'Bsmt Exposure', 'BsmtFin Type 1', 'BsmtFin SF 1', 'BsmtFin Type 2', 'BsmtFin SF 2', 'Bsmt Unf SF', 'Total Bsmt SF', 'Bsmt Full Bath', 'Bsmt Half Bath']

In [ ]:
q1 = '(`Bsmt Qual` == "NA") | (`Bsmt Exposure` == "NA") | (`BsmtFin Type 1` == "NA") | (`BsmtFin Type 2` == "NA")'
q2 = '~((`Bsmt Qual` == "NA") & (`Bsmt Exposure` == "NA") & (`BsmtFin Type 1` == "NA") & (`BsmtFin Type 2` == "NA") & (`BsmtFin SF 1` == 0.00) & (`BsmtFin SF 2` == 0.00) & (`Bsmt Unf SF` == 0.00) & (`Total Bsmt SF` == 0.00) & (`Bsmt Full Bath` == 0) & (`Bsmt Half Bath` == 0))'
df.query(q1).query(q2)[basement_columns]

__Observation__:
- The previous entries show some kind of inconsistency, among the variables describing the basement facilities

__Solution__:
- Since these are only a few entries, we'll fix the entries using common sense
- We'll populate the column `'Bsmt Exposure'` of entries 875, 1681 and 1726, where  indicates there is no basement but we there's a basement in Unfinished conditions, with `'Bsmt Exposure'=='No'` that indicates no exposure to outside

In [ ]:
df.loc[104, ['BsmtFin SF 1','BsmtFin SF 2', 'Bsmt Unf SF', 'Total Bsmt SF', 'Bsmt Full Bath', 'Bsmt Half Bath']] = 0
df.loc[602, ['BsmtFin Type 2']] = 'Unf'
df.loc[[875,1681, 1726], ['Bsmt Exposure']] = 'No'

In [ ]:
df[basement_columns].head()

Next we check if the features representing different surface areas in basements are consistent among them.

In [ ]:
inconsistent_basement_surface = df['BsmtFin SF 1'] + df['BsmtFin SF 2'] + df['Bsmt Unf SF'] != df['Total Bsmt SF']
inconsistent_basement_surface.sum()

__Observation__:
- The previous entries show some kind of inconsistency, among the variables describing the garage facilities

__Solution__:
- Since these are only a few entries, we'll remove the these entries

### Garage Variables

In [39]:
garage_columns = ['Garage Type', 'Garage Finish', 'Garage Cars', 'Garage Area', 'Garage Qual', 'Garage Cond']

In [ ]:
df[garage_columns].head()

In [ ]:
q1 = '(`Garage Type` == "NA") | (`Garage Finish` == "NA") | (`Garage Cars` == 0) | (`Garage Area` == 0) | (`Garage Qual` == "NA") & (`Garage Cond` == "NA")'
q2 = '~((`Garage Type` == "NA") & (`Garage Finish` == "NA") & (`Garage Cars` == 0) & (`Garage Area` == 0) & (`Garage Qual` == "NA") & (`Garage Cond` == "NA"))'
df.query(q1).query(q2)[garage_columns]

In [ ]:
df.drop([1114, 1565], inplace=True)

__Observation__:
- For the case of categorical it normally doesn't make sense to talk about outliers.
- There almost always labels with much less entries but this in principle due to the nature of the feature and the dataset

# Feature distribution:

In [ ]:
n_cols = 5
n_rows = 9

fig, axs = plt.subplots(n_rows, n_cols, figsize=(15, 20))

for i, col in enumerate(categorical_columns):
    r, c = np.divmod(i, n_cols)
    df[col].value_counts().plot(kind='bar', ax=axs[r, c], logy=True)
    axs[r, c].set_title(col, y=1.0)
    axs[r, c].set_xlabel('')

plt.tight_layout()
plt.show()

In [ ]:
n_cols = 5
n_rows = 4

fig, axs = plt.subplots(n_rows, n_cols, figsize=(15, 10))

for i, col in enumerate(discrete_columns):
    r, c = np.divmod(i, n_cols)
    df[col].hist(bins=50, ax=axs[r, c], log=True)
    axs[r, c].set_title(col, y=1.0)
    axs[r, c].set_xlabel('')

plt.tight_layout()
plt.show()

In [ ]:
n_cols = 5
n_rows = 4

fig, axs = plt.subplots(n_rows, n_cols, figsize=(15, 10))

for i, col in enumerate(continuous_columns):
    r, c = np.divmod(i, n_cols)
    df[col].hist(bins=50, ax=axs[r, c], log=True)
    axs[r, c].set_title(col, y=1.0)
    axs[r, c].set_xlabel('')

plt.tight_layout()
plt.show()

# Relationship with target variable

In [ ]:
n_cols = 5
n_rows = 9

fig, axs = plt.subplots(n_rows, n_cols, figsize=(20, 20))

for i, col in enumerate(categorical_columns):
    r, c = np.divmod(i, n_cols)
    sns.boxplot(x=col, y='SalePrice', data=df, ax=axs[r, c])
    #axs[r, c].set_yscale('log')
    axs[r, c].set_title(col, y=1.0)
    axs[r, c].set_xlabel('')
    for label in axs[r, c].get_xticklabels():
        label.set_rotation(45)  # Rotate x-axis labels

plt.tight_layout()
plt.show()

__Observations__:
- There's a big diversity on how the categorical features relate to the 'Sale Price'
- For a few features there's a big effect, depending on the labels and specially for predicting high values of 'Sale Price', e.g. for 'neighbourhood', 'Condition 2', 'Pool Qc'
- For many of the ordinal features there seems to be trend that could be well adjusted by a linear relationship if the labels were ordered differently
- Other features show almost no correlation, e.g. 'Slope', 'Lot Config', 'Lot Shape'
- Additionally, others seem to show no effect with respect with most of the labels in the feature except for 1 or 2, e.g. 'Pool QC'  

In [ ]:
correlations = df[continuous_columns.tolist()].corr()['SalePrice'].abs().sort_values(ascending=False)
print(correlations)

In [ ]:
df_categorical_missing = df[categorical_columns].isna().sum().to_frame(name='total')
df_categorical_missing = df_categorical_missing[df_categorical_missing['total']!=0].sort_values('total', ascending=False)
df_categorical_missing['percentage'] = df_categorical_missing['total']/len(df)*100
df_categorical_missing.head()

# Relationship with target variable

In [ ]:
n_cols = 6
n_rows = 3

fig, axs = plt.subplots(n_rows, n_cols, figsize=(20, 20))

for i, col in enumerate(discrete_columns):
    r, c = np.divmod(i, n_cols)
    sns.scatterplot(x=col, y='SalePrice', data=df, ax=axs[r, c])
    #axs[r, c].set_yscale('log')
    axs[r, c].set_title(col, y=1.0)
    axs[r, c].set_xlabel('')
    for label in axs[r, c].get_xticklabels():
        label.set_rotation(45)  # Rotate x-axis labels

plt.tight_layout()
plt.show()

In [ ]:
print(len(df.columns))
print(len(df[discrete_columns].columns))
print(len(df[continuous_columns].columns))
print(len(df[categorical_columns].columns))

In [ ]:
correlations = df[discrete_columns.tolist()+['SalePrice']].corr()['SalePrice'].abs().sort_values(ascending=False)
print(correlations)

Feature distribution:

In [ ]:

n_cols = 6
n_rows = 3

fig, axs = plt.subplots(n_rows, n_cols, figsize=(15, 10))

for i, col in enumerate(discrete_columns):
    r, c = np.divmod(i, n_cols)
    df[col].hist(bins=50, ax=axs[r, c], log=True)
    axs[r, c].set_title(col, y=1.0)
    axs[r, c].set_xlabel('')
    
plt.tight_layout()
plt.show()

# Enconding of Categorical Variables

__Observation__:
- Features such as 'Exter Cond', 'Exter Cond', 'Garage Finish', have an ordinal nature despite being part of the categorical variables
- We can use this fact to directly encode them into a single other numerical discrete variable

  
__Solution__:
- Encode the following categorical ordinal features: 'Lot Shape', 'Utilities', 'Land Slope', 'Overall Qual', 'Overall Cond', 'Exter Qual', 'Exter Cond', 'Bsmt Qual', 'Bsmt Cond', 'Bsmt Exposure', 'BsmtFin Type 1', 'BsmtFin Type 2', 'HeatingQC', 'Electrical', 'KitchenQual', 'Functional', 'FireplaceQu', 'Garage Finish', 'Garage Qual', 'Garage Cond', 'Paved Drive'
- Assume a linear relationship between 'SalePrice' and all the categorical ordinal features (following observation of previous plots and information in documentation)
- Label encode these categorical variables into numerical discrete ones,  avoiding the creation of many new extra variables with "one-hot encoding"

In [ ]:
encoding_order={
    'Lot Shape':['Reg', 'IR1', 'IR2', 'IR3'], 
    'Utilities':['AllPub', 'NoSewr', 'NoSeWa', 'ELO'],
    'Land Slope':['Gtl', 'Mod', 'Sev'],
    'Overall Qual': ['10', '9', '8', '7', '6', '5', '4', '3', '2', '1'],
    'Overall Cond': ['10', '9', '8', '7', '6', '5', '4', '3', '2', '1'],
    'Exter Qual':['Ex', 'Gd', 'TA', 'Fa', 'Po'],
    'Exter Cond':['Ex', 'Gd', 'TA', 'Fa', 'Po'],
    'Bsmt Qual':['Ex', 'Gd', 'TA', 'Fa', 'Po', 'NA'],
    'Bsmt Cond':['Ex', 'Gd', 'TA', 'Fa', 'Po', 'NA'],
    'Bsmt Exposure':['Gd', 'Av', 'Mn', 'No', 'NA'], 
    'BsmtFin Type 1':['GLQ', 'ALQ', 'BLQ', 'Rec', 'LwQ', 'Unf', 'NA'],
    'BsmtFin Type 2':['GLQ', 'ALQ', 'BLQ', 'Rec', 'LwQ', 'Unf', 'NA'],
    'Heating QC':['Ex', 'Gd', 'TA', 'Fa', 'Po'],
    'Electrical':['SBrkr', 'FuseA', 'FuseF', 'FuseP', 'Mix'],
    'Kitchen Qual':['Ex', 'Gd', 'TA', 'Fa', 'Po'],
    'Functional':['Typ', 'Min1', 'Min2', 'Mod', 'Maj1', 'Maj2', 'Sev', 'Sal'],
    'Fireplace Qu':['Ex', 'Gd', 'TA', 'Fa', 'Po', 'NA'],
    'Garage Finish':['Fin', 'RFn', 'Unf'],
    'Garage Qual':['Ex', 'Gd', 'TA', 'Fa', 'Po', 'NA'],
    'Garage Cond':['Ex', 'Gd', 'TA', 'Fa', 'Po', 'NA'],
    'Paved Drive':['Y', 'P', 'N'],
    'Pool QC': ['Ex', 'Gd', 'TA', 'Fa', 'NA'],
    'Fence': ['GdPrv', 'MnPrv', 'GdWo', 'MnWw', 'NA'],
    }

In [ ]:
print("Missing values: ")
for col in encoding_order.keys():
    print(f'{col:20}: {df[col].isna().sum().sum()}')

In [ ]:
from sklearn.preprocessing import OrdinalEncoder


ordinal_encoder = OrdinalEncoder(categories=[['bad', 'average', 'good']])
df['col'] = ordinal_encoder.fit_transform(df[['col']])

# EDA and Data Cleaning

### Missing values
We start by observing the occurrence of missing values and next we see how to address these

In [ ]:
df.isnull().sum().sum()

In [ ]:
df_missing = df.isna().sum().to_frame(name='total')
df_missing = df_missing[df_missing['total']!=0].sort_values('total', ascending=False)
df_missing['percentage'] = df_missing['total']/len(df)*100
df_missing

In [ ]:
df.isnull().sum().sum()

In [ ]:
df_missing = df.isna().sum().to_frame(name='total')
df_missing = df_missing[df_missing['total']!=0].sort_values('total', ascending=False)
df_missing['percentage'] = df_missing['total']/len(df)*100
df_missing

Except for 'Electrical', all reamining columns with missing values are numerical. So let's handle 'Electrical' first. We chose the most common value to fill in the only missing entry.

In [ ]:
df['Electrical'].fillna(value=df['Electrical'].mode, inplace=True)

Reading the description of 'Lot Frontage' one sees that this is a continuous variable indicating 'Linear feet of street connected to property', it seems reasonable to assume if a house has no lot frontage then this could be set to 0.

In [ ]:
df['Lot Frontage'].hist(bins=50)
df['Lot Frontage'].fillna(0, inplace=True)

We apply the same principle to all other missing values, which at this point are present only in numerical columns.

In [ ]:
## CHECK!!
df.fillna(0, inplace=True)

In [ ]:
df.isnull().sum().sum()

### Inconsistencies

In [ ]:
sum(df['Year Remod/Add'] < df['Year Built'])

In [ ]:
(df['Bsmt Qual']=='None')!=(df['Bsmt Cond']=='None')

In [ ]:
sum((df['Bsmt Qual']=='None')+(df['Bsmt Cond']=='None'))

In [ ]:
print(sum(df['Bsmt Qual']=='None'))
print(sum(df['Bsmt Cond']=='None'))
print(sum(df['Bsmt Exposure']=='None'))
print(sum((df['Bsmt Qual']=='None') & (df['Bsmt Cond']=='None')))


In [ ]:
((df['Bsmt Qual']=='None')!=(df['Bsmt Cond']=='None')+(df['Bsmt Exposure']=='None'))

### Outliers

__Exploring the categorical variables__

In [ ]:
df_cat = df.loc[:,df.dtypes == 'object'].copy()
df_cat.nunique()

__Exploring the numerical variables__

In [ ]:
df.describe().loc[['min', 'max']].T

__Observation:__
- 'Garage Yr Blt': Minimum value is 0 and maximum is 2207 which clearly is an error since dataset only goes up to 2010

__Solution:__
- Remove entries where 'Garage Yr Blt' > 2010
- Remove entries where 'Garage Yr Blt' > 0

In [ ]:
mask = (df['Garage Yr Blt'] > 0) & (df['Garage Yr Blt'] < 2010)
df = df[mask]
df.shape

In [ ]:
df.hist(bins=20, figsize=(20,20), log=True)
plt.show()

In [ ]:
pd.set_option('display.float_format', '{:.5f}'.format)
corr_series = df[cols_per_hundred].corrwith(df['energy_per_hundred'])
corr_series.sort_values(ascending=False)